torch/distributed/_tensor/experimental/dmap.py

```diff
@@ -0,0 +1,155 @@
+        >>> dX = distribute_tensor(X, device_mesh, row_wise)  # row-wisely sharded X tensor
+        >>> dY = d_mm_allreduce_forward(dW, dX)  # apply d_mm_allreduce_forward to DTensors
+
+    NOTE: This API is currently experimental and subject to change
+    """
+
+    def wrapped(*args, **kwargs):
+        # process input args
+        flat_args, args_spec = tree_flatten(args)
+
+        # we assume every DTensor object is placed on the same device mesh
+        flat_local_args = []
+        nonlocal device_mesh  # access var device_mesh from the outer scope
+        for idx, arg in enumerate(flat_args):
+            if isinstance(arg, DTensor):
+                if device_mesh is None:  # infer device mesh from the DTensor arg
+                    device_mesh = arg.device_mesh
+
+                assert arg.device_mesh == device_mesh, (
+                    f"arg {arg} in dmap has a mismatched device mesh:"
+                    f"{arg} has device mesh {arg.device_mesh} while"
+                    f"the input device mesh is {device_mesh}!"
+                )
+                if in_placements is not None:
+                    spec = (
+                        in_placements[idx]
+                        if isinstance(in_placements, tuple)
+                        else in_placements
+                    )
+                    if not isinstance(spec, tuple):
+                        spec = tuple(spec)
+
+                    if arg.placements != spec:
+                        if resharding:
+                            # redistribute to input placements
+                            arg = arg.redistribute(device_mesh, spec)
+                        else:
+                            raise ValueError(
+                                f"arg {arg} in dmap has a mismatched placements:"
+                                f"arg placements is {arg.placements} but the input"
+                                f"placements is {spec}!"
+                                "If resharding is wanted, set resharding=True to dmap."
+                            )
+
+                flat_local_args.append(arg.to_local())
+            else:
+                flat_local_args.append(arg)
+
+        local_args = tree_unflatten(flat_local_args, args_spec)
+
+        out = func(device_mesh, *local_args, **kwargs)
+
+        # process output
+        flat_out, out_spec = tree_flatten(out)
+        flat_dist_out: List[Optional[DTensor]] = []
+        for idx, out in enumerate(flat_out):
+            spec = (
+                out_placements[idx]
+                if isinstance(out_placements, tuple)
+                else out_placements
+            )
+            if isinstance(out, torch.Tensor):
+                assert not isinstance(
+                    out, DTensor
+                ), f"torch.Tensor output expected but received {type(out)}: {out}"
+
+                flat_dist_out.append(DTensor.from_local(out, device_mesh, spec))
```

Comment:
Good question! Generally in dtensor we don't check the type of `spec` because in Python without individually checking every element we can only assert if `spec` itself is a `Sequence`. So usually we just check if `spec` is `None` or not.